main/src/ca/uwaterloo/flix/language/phase/Dependencies.scala

```diff
@@ -1,88 +1,719 @@
-import ca.uwaterloo.flix.api.lsp.acceptors.AllAcceptor
-import ca.uwaterloo.flix.language.ast.{SourceLocation, Type, TypeConstructor, TypedAst}
-import ca.uwaterloo.flix.language.ast.TypedAst.Root
-import ca.uwaterloo.flix.language.ast.shared.{DependencyGraph, Input, SymUse}
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Record
+import ca.uwaterloo.flix.language.ast.TypedAst.Predicate.Body
+import ca.uwaterloo.flix.language.ast.{Scheme, SourceLocation, Type, TypeConstructor, TypedAst}
+import ca.uwaterloo.flix.language.ast.TypedAst.{Expr, Pattern, RestrictableChoosePattern, Root}
+import ca.uwaterloo.flix.language.ast.shared.{BroadEqualityConstraint, DependencyGraph, EqualityConstraint, Input, SymUse, TraitConstraint}
 import ca.uwaterloo.flix.language.dbg.AstPrinter.*
+import ca.uwaterloo.flix.util.ParOps
 import ca.uwaterloo.flix.util.collection.MultiMap
 
-object Dependencies {
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentMap}
 
-  /** Checks the safety and well-formedness of `root`. */
+object Dependencies {
+  /**
+    * Computes the dependency graph of the given `root`.
+    *
+    * Most of the time we just recursively visit the sub-nodes. We only add dependencies for three base cases:
+    *   - Type
+    *   - SymUse
+    *   - Instance
+    */
   def run(root: Root)(implicit flix: Flix): (Root, Unit) = flix.phaseNew("Dependencies") {
+    implicit val sctx: SharedContext = SharedContext(new ConcurrentHashMap[(Input, Input), Unit]())
+    ParOps.parMap(root.defs.values)(visitDef)
+    ParOps.parMap(root.effects.values)(visitEff)
+    ParOps.parMap(root.enums.values)(visitEnum)
+    ParOps.parMap(root.instances.values)(visitInstances)
+    ParOps.parMap(root.structs.values)(visitStruct)
+    ParOps.parMap(root.traits.values)(visitTrait)
+    ParOps.parMap(root.typeAliases.values)(visitTypeAlias)
 
-    // TODO: We should not depend on Consumer from LSP. Instead we should traverse the AST manually.
-    // Moreover, we should traverse the AST in parallel and using changeSet.
+    var deps = MultiMap.empty[Input, Input]
+    sctx.deps.forEach((k, _) => deps = deps + k)
+    val dg = DependencyGraph(deps)
+    (root.copy(dependencyGraph = dg), ())
+  }
 
-    object consumer extends Consumer {
-      var deps: MultiMap[Input, Input] = MultiMap.empty
+  /**
+    * Adds a dependency between the source and destination locations.
+    * The value is fixed to () since it doesn't matter.
+    */
+  private def addDependency(src: SourceLocation, dst: SourceLocation)(implicit sctx: SharedContext): Unit = {
+    sctx.deps.put((src.sp1.source.input, dst.sp1.source.input), ())
+  }
 
-      /**
-        * Adds a dependency `src -> dst` signifying that if `src` changes then `dst` must be recomputed.
-        */
-      private def addDependency(src: SourceLocation, dst: SourceLocation): Unit = {
-        deps += (src.sp1.source.input, dst.sp1.source.input)
-      }
+  /**
+    * Returns the dependencies of the given type.
+    */
```

Comment:
The order of the file is a bit wonky. Here we have visitType as the first visit function.

Here I would expect these functions: visitDef, visitEff, visitEnum.... to follow the order used in def run.

Then after those we have visitExpr and visitType

and then later we have various helper functions.

We don't have an exact order, but something like the above. You can see how it looks in e.g. Safety.scala - although it might not be perfect.